src/util/triplea/MapXMLCreator/CanalDefinitionsPanel.java

```diff
@@ -0,0 +1,442 @@
+      if (canalTerrs == null) {
+        canalTerrs = Tuple.of(new TreeSet<String>(), new TreeSet<String>());
+        MapXMLHelper.putCanalDefinitions(currentCanalName, canalTerrs);
+      }
+
+      selectedWaterTerritories = canalTerrs.getFirst();
+      selectedLandTerritories = canalTerrs.getSecond();
+
+      final SortedSet<String> selectedTerrsSameType;
+      if (newTerrIsWater)
+        selectedTerrsSameType = selectedWaterTerritories;
+      else
+        selectedTerrsSameType = selectedLandTerritories;
+
+      if (selectedTerrsSameType.size() > 0) {
+        final SortedSet<String> commonNeighborsDiffType =
+            getCommonNeighborsOfType(selectedTerrsSameType, !newTerrIsWater);
+        commonNeighborsDiffType.retainAll(newTerrNeighborsDiffType);
+
+        if (commonNeighborsDiffType.size() < 2) {
+          JOptionPane.showMessageDialog(null, "The selected " + (newTerrIsWater ? "water" : "land")
+              + " territory is connected to less than 2 common " + (!newTerrIsWater ? "water" : "land")
+              + " territories with the other " + (newTerrIsWater ? "water" : "land") + " territor"
+              + (selectedTerrsSameType.size() == 1 ? "y" : "ies")
+              + "!\rRight click to deselect current canal '" + currentCanalName + "'.", "Input Error",
+              JOptionPane.ERROR_MESSAGE);
+          return;
+        }
+      }
+
+      if (selectedTerrsSameType.contains(newTerrName))
+        selectedTerrsSameType.remove(newTerrName);
+      else
+        selectedTerrsSameType.add(newTerrName);
+
+    }
+
+    SwingUtilities.invokeLater(new Runnable() {
+      public void run() {
+        imagePanel.repaint();
+      }
+    });
+  }
+
+  protected void clearSelection() {
+    currentCanalName = null;
+    selectedLandTerritories = new TreeSet<String>();
+    selectedWaterTerritories = new TreeSet<String>();
+  }
+
+  protected LinkedHashSet<String> getNeighborsByType(final String newTerrName, final boolean waterNeighbors) {
+    final LinkedHashSet<String> neighborsByType = new LinkedHashSet<String>();
+    final LinkedHashSet<String> neighbors = MapXMLHelper.territoryConnections.get(newTerrName);
+    for (final Entry<String, LinkedHashSet<String>> terrConnEntry : MapXMLHelper.territoryConnections.entrySet()) {
+      if (MapXMLHelper.territoryDefintions.get(terrConnEntry.getKey()).get(DEFINITION.IS_WATER) == waterNeighbors
+          && terrConnEntry.getValue().contains(newTerrName))
+        neighborsByType.add(terrConnEntry.getKey());
+    }
+    if (neighbors != null) {
+      for (final String neighbor : neighbors) {
+        if (MapXMLHelper.territoryDefintions.get(neighbor).get(DEFINITION.IS_WATER) == waterNeighbors)
+          neighborsByType.add(neighbor);
+      }
+    }
+    return neighborsByType;
+  }
+
+  protected SortedSet<String> getCommonNeighborsOfType(final SortedSet<String> terrList, final boolean typeIsWater) {

```

Comment:
Another way to avoid boolean parameters:
```
 protected SortedSet<String> getNeighborWaterTerritories(final SortedSet<String> terrList) {
   return this.getCommonNeighborsOfType(terrList, true );
}

 protected SortedSet<String> getNeighborLandTerritories(final SortedSet<String> terrList) {
   return this.getCommonNeighborsOfType(terrList, false);
}

 private SortedSet<String> getCommonNeighborsOfType(final SortedSet<String> terrList, final boolean typeIsWater) {
            :
:
```
Now compare how the client code would look:
```
getNeighborWaterTerritories(terrs)
getNeighborLandTerritories(terrs)
```
vs
```
getCommonNeighborsOfType(terrs, true);
getCommonNeighborsOfType(terrs, false)
```